enonic/src/main/resources/site/parts/article-form/article-form.ts

```diff
@@ -47,9 +52,31 @@ function createThymeleafParams(res: QueryResponse<Comment>,article: Content<Arti
   return {
     title: article.data.title,
     body: article.data.body,
-    comments: res.hits.map((content: Content<Comment>) => content.data),
+    comments: createCommentTree(res.hits),
     serviceUrl: serviceUrl('comment'),
     parentPath: article._path,
     pathSelf: path
   }
+}
+
+const byParentFirst = pipe(ordBoolean, contramap((c: Content<Comment>) => !!c.data.parentId))
+
+export function childrenToParent (acc: ContentTree<Comment>[], curr: Content<Comment>): ContentTree<Comment>[] {
+  if(curr.data.parentId) {
+    const parent = acc.filter((item) => item._id === curr.data.parentId);
+    if(parent[0]) { // only add child when there is a parent (dont show reply to comments that have been deleted)
+      const children = parent[0]?.children ?? []
+      parent[0].children = [...children, curr]
+    }
+    return [...acc]
+  } else {
+    return [...acc, curr]
+  }
+}
+
+interface ContentTree<T> extends Content {
+  children?: Content[]
+}
+export function createCommentTree (list: ReadonlyArray<Content<Comment>>):ContentTree<Comment>[] {
+  return pipe(sort(byParentFirst)([...list]), reduce([], childrenToParent))
```

Comment:
I don't think the defensive copy is necessary here. I think `sort` from *fp-ts* doesn't mutate the input (it wouldn't be the functional way.).

Reply:
Prefer to stack your pipes vertically. You can also pipe the `list` into  `sort`, it might be slightly cleaner.:
```suggestion
  return pipe(
    list,
    sort(byParentFirst), 
    reduce([], childrenToParent)
  )
```

Reply:
An other possible choice here (since `list`was the first parameter in the pipe), is to use more classic composition with `flow`, which only combines functions to a new function, without passing in a value first. This is exactly the same.

```typescript
const createCommentTree = flow(
  sort(byParentFirst), 
  reduce([], childrenToParent), 
)
```